llvm/docs/GitHub.rst

```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
 
 Pull Requests
 =============
-The LLVM Project does not currently accept pull requests for the llvm/llvm-project
-repository.  However, there is a
-`plan <https://discourse.llvm.org/t/code-review-process-update/63964>`_ to move
-to pull requests in the future.  This section documents the pull request
-policies LLVM will be adopting once the project starts using them.
+The LLVM project is using GitHub Pull Requests for Code Reviews. This document
+describes the typical workflow of creating a Pull Request and getting it reviewed
+and accepted. This is meant as an overview of the GitHub workflow, for compelete
+documentation refer to `GitHubs documentation <https://docs.github.com/pull-requests>`_.
+
+GitHub Tools
+------------
+You can interact with GitHub in several ways: via git command line tools,
+the web browser, `GitHub Desktop <https://desktop.github.com/>`_, or the
+`GitHub CLI <https://cli.github.com>`_. This guide will cover the git command line
+tools and the GitHub CLI. The GitHub CLI (`gh`) will be most like the `arc` workflow and
+recommended.
 
 Creating Pull Requests
-^^^^^^^^^^^^^^^^^^^^^^
+----------------------
 For pull requests, please push a branch to your
 `fork <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks>`_
 of the llvm-project and
 `create a pull request from the fork <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork>`_.
 
+Creating Pull Requests with GitHub CLI
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+With the CLI it's enough to create the branch locally and then run:
```

Comment:
Should we be explicit that the original PR creation *must always* have a single commit?
If our advice is to squash and merge, there is an implication here that all PRs must start as a single commit.

This is the scenario I'm worried about: a well intentioned developer creates a lot of small patches (great!) but make a single PR, because they are used to other GitHub flows. The current advice will destroy all the nice work done to create the small commits.

Reply:
Yes, this is also my concern. For this reason, I strongly prefer rebasing and squashing in fixes and forcepushing for each update of the branch. I know github has been rather bad at tracking old review comments across force pushes before, but I have a faint feeling it's working better these days. (And if we are going to squash at the end, or if we do allow the occasional rebase anyway when needed, we will lose context anyway.)

Personally, when submitting a PR, I would be very vocal about this, about how I want it merged (squashed, or kept as is, or if we insist on incremental fixes in separate commits, I would say that I want to do the final squashing once things are approved, and only then let someone merge it, with rebase+merge instead of squash+merge).

Reply:
See my comment below, I think there is a valid discussion to have here, but not sure this is the right place to have it. We need to get the documentation in, and what's described here is the **closest** thing to the consensus I have found.

Reply:
Just to make sure we are on the same page, I am not proposing anything new or opposing the squash and merge flow; this is what the majority of people involved in the discussion agreed upon.
The rationale for that decision (AFAICT) was to mimic the phabricator workflow as closely as possible, and that workflow starts on the premise that *one* unit of contribution is reviewed through a Differential review. To emphasize this intent, I do believe that our new documentation should state that "PRs should be created from a single commit", it is a 1:1 match to what exists in phabricator today.

Reply:
How about adding this:

```
Keep in mind that each pull request should generally only contain one commit.
This makes it easier for reviewers to understand the introduced changes and
provide feedback. It also helps maintain a clear and organized commit history
for the project. If you have multiple changes you want to introduce, it's
recommended to create separate pull requests for each change.
```

Reply:
Looks great! :)

Reply:
Note we also have a section that advise against amend+force-push for the PR workflow ("Updating Pull-request" below)

Reply:
We currently only allow squash and merge for pull requests, I think this is a good policy because it makes it harder to make a mistake and accidentally push a series of fixup commits.  Even with this setting though, it's still possible to have a pull request with multiple commits from the start, you just have to push it manually once it's reviewed rather than going through the PR interface.

Reply:
> , you just have to push it manually once it's reviewed rather than going through the PR interface.

This is a great point and I am glad it is a possibility, but it seems like an advanced use for the following reasons:

* This is likely not what most GH users are used to doing, so I think the current wording is appropriate.
* GH will not close the review automatically and there won't be any "closed by commit X" messages like we get in phab. Manual intervention is required here.
*  This doesn't play well with the advice of not force-pushing (we usually want to address comments *on the commit* they were mentioned, so that the log is preserved properly after merge).
* CI won't be testing each commit to ensure they all compile / test.